src/nxt_array.h

```diff
@@ -9,34 +10,32 @@
 
 
 typedef struct {
-    void      *elts;
-    /* nelts has uint32_t type because it is used most often. */
-    uint32_t  nelts;
-    uint16_t  size;
-    uint16_t  nalloc;
-    nxt_mp_t  *mem_pool;
+    void                *elts;
+    nxt_uint_t          nelts;
+    nxt_uint_t          size;
+    nxt_uint_t          nalloc;
+    nxt_mp_t            *mem_pool;
 } nxt_array_t;
 
 
-nxt_inline void
```

Comment:
removed incorrect array initialization (nxt_array_init) with a pointer to a non-existent set of elements;

Reply:
It's non existing in the particular `nxt_var_query_init()` case, but it may exist in other cases that this function in designed for (but they aren't yet represented in the code).

There's nothing incorrect in setting pointer to the end of array structure as soon as the correct number of initialized elements is provided (zero in that case).

The only issues with this function is that it's missing the "nalloc" parameter to be used in non-zero cases.

Reply:
> It's non existing in the particular `nxt_var_query_init()` case, but it may exist in other cases that this function in designed for (but they aren't yet represented in the code).

I already made the decision to revert `nxt_array_init` and discard changes to **nxt_var.c** back in @hongzhidao answer. I will also take into account additional information about possible upcoming changes that you provided. Thanks.

> 
> There's nothing incorrect in setting pointer to the end of array structure as soon as the correct number of initialized elements is provided (zero in that case).

So you don't see any problem with using `nxt_array_init` with `array->elts = nxt_pointer_to(array, sizeof(nxt_array_t));` in the context of `nxt_var_query_s`:
- `values` points to `parts`
- `parts` points to `cache`
at initialization time? 

Instead of being `NULL` for example until `array->elts` is fully initialized with `nxt_arrray_add` for example?

> 
> The only issues with this function is that it's missing the "nalloc" parameter to be used in non-zero cases.

I took it into account, even at the moment when @hongzhidao answered. But thanks for pointing it out to me.

Reply:
> Instead of being NULL for example until array->elts is fully initialized with nxt_array_add for example?

The "array->elts" serves for two different purposes:
1. It indicates whenever elements were allocated from the same memory as the array structure. In this case this memory can't be freed.
2. It points to the allocated elements, if any.

By setting it to NULL in the current code, you will break the first purpose, so the check in nxt_array_add():
```
        if (array->elts != nxt_pointer_to(array, sizeof(nxt_array_t))) {
            nxt_mp_free(array->mem_pool, array->elts);
        }
```
won't work right.

Reply:
Perhaps I expressed myself incorrectly. I didn't suggest using NULL with the old code. I think it's obvious to everyone where this will lead.

Sorry, but it seems to me that you somehow selectively answer the questions I asked. Pulling out of the context of my question only what is necessary for you in order to make it meaningless or incorrect.

In the future, I will try to explain the essence of my ideas (for example) in more detail, so that you do not have the opportunity to double-read about the questions I asked and the opportunity, as it seems to me, to present them in the way you need.

**I remind you that you still haven't answered the question:**
>
> So you don't see any problem with using `nxt_array_init` with `array->elts = nxt_pointer_to(array, sizeof(nxt_array_t));` in the context of `nxt_var_query_s`:

> `values` points to `parts`
> `parts` points to `cache`
> at initialization time?

Reply:
> So you don't see any problem with using nxt_array_init with array->elts = nxt_pointer_to(array, sizeof(nxt_array_t)); in the context of nxt_var_query_s:

As long as this pointer works only like a flag (and it works like a flag, which I've tried to explain by naming two purposes of the pointer) I see no problem here.

> values points to parts
> parts points to cache
> at initialization time?

"values" doesn't point to "parts".  It points to the end of the array structure in "nxt_var_query_s", where zero number of elements should start.  Sure, due to zero number of elements this value is equal to position of the "parts" field.  Which is just a coincidence in this particular case.  But, for example, if there would be not zero, but N elements before the "parts" field, how is this different?

If it overflows, it will corrupt the "parts" and possibly other fields. If it will be placed somewhere else, then it may corrupt some other structures - no difference here too.

In your version of code you've allocated a space from memory pool for 1 element.

So I don't see how overflowing 0 elements is different from overflowing 1 element (2, or 3, or N... doesn't matter). And I don't see how the consequences of overwriting the "nxt_var_query_s" structure is more dangerous to any consequences of overflowing any other structure that can be placed in the memory pool right after the 1 allocated element in your code.

Reply:
I'm sorry, I made a mistake with the names:
`values.elts` (offset size - `nxt_array_t`) - the actual pointer goes to `parts`;
`parts.elts` (offset size - `nxt_array_t`) - actual pointer goes to `cache`;
It's about the old `nxt_array_init` code.

Regarding the changes affecting **nxt_var.c**, I refused as soon as @hongzhidao drew my attention to the design, as I understand that they gave an insufficient assessment of the correctness of the work of all components.

The difference is that the initialization of the `array->elts` object that I propose and the allocation of memory for it are initially independent of the main array structure and the memory allocated for its storage. Which is allocated and released as the array grows without leaving traces in the form of abandoned memory areas in favor of 1 additional call when creating the array. The array structure in itself contains only a pointer to the set of elements stored in it, which is replaced without any problems by another pointer, while guaranteed to free the previously used memory area.

The question here is not only about possible overflow and overwriting, since it can be achieved in other simpler ways if desired. And that the pointer is initially set to a place in memory that has nothing to do with the current set of elements. In my understanding, if it does not exist, then it should be NULL, and not point to another object. I remember you said that the worst thing is if the developers do not understand why the code was written and for what end goal it was done. Here is the same case already contained in the current version of the code, which is sincerely incomprehensible to me and actually misleads me. This piece of code looks like it was copied from nxt_array_create and the programmer simply forgot about determining the dimension of the array, considering that this pointer refers to an existing list of elements.

Reply:
> The difference is that the initialization of the array->elts object that I propose and the allocation of memory for it are initially independent of the main array structure and the memory allocated for its storage.

The initial design was made this way because in many cases requests are processed without any variables interpolation using only const string values. So, the `nxt_var_query_init()` call allocates the query structure, then `nxt_var_query()` calls actually do nothing as they return on checking `nxt_var_is_const(var)`, so we have just one allocation in this common case. If there will be variable strings, then we will have 3 allocations in total (query structure and two array elements).

After your change, there will be always 3 allocations.

> This piece of code looks like it was copied from nxt_array_create and the programmer simply forgot about determining the dimension of the array, thinking that this pointer refers to an existing list of elements.

You can improve it this way to make it look more consistent:
```c
nxt_inline void
nxt_array_init(nxt_array_t *array, nxt_mp_t *mp, nxt_uint_t n, size_t size)
{
    array->elts = nxt_pointer_to(array, sizeof(nxt_array_t));
    array->nelts = 0;
    array->size = size;
    array->nalloc = n;
    array->mem_pool = mp;
}


nxt_array_t *
nxt_array_create(nxt_mp_t *mp, nxt_uint_t n, size_t size)
{
    nxt_array_t  *array;

    array = nxt_mp_alloc(mp, sizeof(nxt_array_t) + n * size);

    if (nxt_slow_path(array == NULL)) {
        return NULL;
    }

    nxt_array_init(array, mp, n, size);

    return array;
}
```
with the relevant changes to `nxt_var.c`:
```diff
diff --git a/src/nxt_var.c b/src/nxt_var.c
--- a/src/nxt_var.c
+++ b/src/nxt_var.c
@@ -433,8 +433,8 @@ nxt_var_query_init(nxt_var_query_t **que
             return NXT_ERROR;
         }
 
-        nxt_array_init(&query->values, mp, sizeof(nxt_var_value_t));
-        nxt_array_init(&query->parts, mp, sizeof(nxt_str_t *));
+        nxt_array_init(&query->values, mp, 0, sizeof(nxt_var_value_t));
+        nxt_array_init(&query->parts, mp, 0, sizeof(nxt_str_t *));
 
     } else {
         nxt_array_reset(&query->values);
```

Reply:
It refers to the "existing" zero number of elements. It's also possible to initialize with more than zero number of elements, but `nxt_var_query_s` will be bigger:
```c
struct nxt_var_query_s {
    nxt_array_t         values;   /* of nxt_var_value_t */
    nxt_var_value_t     values_elts[4];
    nxt_array_t         parts;    /* of nxt_str_t * */
    nxt_str_t           parts_elts[4];

    nxt_lvlhsh_t        cache;

    nxt_str_t           *spare;
    nxt_uint_t          waiting;
    nxt_uint_t          failed;   /* 1 bit */

    void                *ctx;
    void                *data;

    nxt_work_handler_t  ready;
    nxt_work_handler_t  error;
};
```

Reply:
I understand why `nxt_array_init` appeared as soon as @hongzhidao pointed out to me the design used in **nxt_var.c**, so I immediately abandoned almost any changes associated with this file, except for the explicit initialization of `nalloc` and `nelts` in `nxt_array_init` and a pointer at a given offset actually setting a pointer to an adjacent object that you define as necessary before the first expansion of the array, whose function is to enforce the bounds of the array. I understand why `nxt_array_init` should be lightweight in the context of a query and why it was originally marked up as inline, so I have no questions about its necessity.